docs/historical/mh-jun-1982/man/file.1
.TH FILE 1 RAND
.SH NAME
file \- file message(s) in (an)other folder(s)
.SH SYNOPSIS
\fBfile \%[\fB\-\fRsrc\ +folder] \%[msgs] \%[\fB\-link\fR] \%[\fB\-preserve\fR] \%+folder\ ...
\%[\fB\-nolink\fR] \%[\fB\-nopreserve\fR]
\%[\fB\-file\fR\ file] \%[\fB\-nofile\fR] \%[\fB\-help\fR]
.SH DESCRIPTION
\fIFile\fR moves (\fImv\fR(1)) or links (\fIln\fR(1)) messages from a
source folder into one or more destination folders.
If you think
of a message as a sheet of paper, this operation is not
unlike filing the sheet of paper (or copies) in file cabinet
folders.
When a message is filed, it is linked into the
destination folder(s) if possible, and is copied otherwise.
As long
as the destination folders are all on the same file system, multiple filing
causes little storage overhead.
This facility provides a good way to cross-file or multiply-index
messages.
For example, if a message is received from Jones about
the ARPA Map Project, the command

     file\0cur\0+jones\0+Map

would allow the message to be found in either of the two
folders `jones' or `Map'.

The option `\-file file' directs \fIfile\fR to use the specified
file as the source message to be filed, rather than a message from
a folder.

If a destination folder doesn't exist, \fIfile\fR will ask if you
want to create one.
A negative response will abort the file
operation.

`\-link' preserves the source folder copy of the message
(i.e., it does a \fIln\fR(1) rather than a \fImv\fR(1)), whereas,
`\-nolink' deletes the ``filed'' messages from the source
folder.
Normally, when a message is filed, it is assigned the
next highest number available in each of the destination folders.
Use of the `\-preserve' switch will override this message
``renaming'', but name conflicts may occur, so
use this switch cautiously.
(See \fIpick\fR for more details on
message numbering.)

If `\-link' is not specified (or `\-nolink' is specified),
the filed messages will be removed (unlink(2)) from the
source folder.
.SH IMPORTANT NOTE
"File" already exists in many Unix systems as a program to check
the type of a file.  If the Rand MH system has been installed,
that program will have been renamed to
.I filetype(1).
.SH FILES
.ta 2.4i
$HOME/\*.mh\(ruprofile	The user profile
.SH PROFILE COMPONENTS
.br
Path:	To determine the user's MH directory
.br
Current-Folder:	To find the default current folder
.br
Folder\-Protect:	To set mode when creating a new folder
.SH SEE ALSO
show(1), folder(1)
.SH DEFAULTS
`\-src +folder' defaults to the current folder
.br
`msgs' defaults to cur
.br
`\-nolink'
.br
`\-nopreserve'
.br
`\-nofile'
.SH CONTEXT
If `\-src +folder' is given, it will become the
current folder for future MH commands.
If neither `\-link' nor
`all' are specified, the current message in the source
folder will be set to the last message specified; otherwise, the
current message won't be changed.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

docs/historical/mh-jun-1982/progs/file.c
#ifdef COMMENT
	Proprietary Rand Corporation, 1981.
	Further distribution of this software
	subject to the terms of the Rand
	license agreement.
#endif

#include "../mh.h"
#include "../folder.h"
#include <stdio.h>

#define NFOLD 20                /* Allow 20 folder specs */
char *calloc();

/* file [-src folder] [msgs] +folder [+folder ...]
 *
 * moves messages from src folder (or current) to other one(s).
 *
 *  all = 1-999 (MAXFOLDER) for a message sequence
 *  -preserve says preserve msg numbers
 *  -link says don't delete old msg
 */

extern struct swit anoyes[];    /* Std no/yes gans array */

int vecp, foldp, prsrvf;
char **vec, maildir[128], *folder;
struct msgs *mp;

struct st_fold folders[NFOLD];

char   *files[NFOLD + 1];       /* Vec of files to process--starts at 1! */
int     filec = 1;

struct swit switches[] = {
	"all",           -3,      /* 0 */
	"link",           0,      /* 1 */
	"nolink",         0,      /* 2 */
	"preserve",       0,      /* 3 */
	"nopreserve",     0,      /* 4 */
	"src +folder",    0,      /* 5 */
	"file",           0,      /* 6 */
	"help",           4,      /* 7 */
	0,                0
};

/*ARGSUSED*/
main(argc, argv)
char *argv[];
{
	register int i, msgnum;
	register char *cp;
	char *msgs[128];
	int msgp, linkf;
	char **ap;
	char *arguments[50], **argp;

	invo_name = argv[0];
#ifdef NEWS
	m_news();
#endif
	folder = 0; msgp = 0; linkf = 0;
	cp = r1bindex(argv[0], '/');
	if((cp = m_find(cp)) != NULL) {
		ap = brkstring(cp = getcpy(cp), " ", "\n");
		ap = copyip(ap, arguments);
	} else
		ap = arguments;
	VOID copyip(argv+1, ap);
	argp = arguments;
	while(cp = *argp++) {
		if(*cp == '-')
			switch(smatch(++cp, switches)) {
			case -2:ambigsw(cp, switches);       /* ambiguous */
				goto leave;
							/* unknown */
			case -1:fprintf(stderr, "file: -%s unknown\n", cp);
				goto leave;
						       /* -all */
			case 0: fprintf(stderr, "\"-all\" changed to \"all\"\n");
				goto leave;
			case 1: linkf = 1;  continue;  /* -link */
			case 2: linkf = 0;  continue;  /* -nolink */
			case 3: prsrvf = 1;  continue; /* -preserve */
			case 4: prsrvf = 0;  continue; /* -nopreserve */
			case 5: if(folder) {           /* -src */
					fprintf(stderr, "Only one src folder.\n");
					goto leave;
				}
				if(!(folder = *argp++) || *folder == '-') {
missing:        fprintf(stderr, "file: Missing argument for %s switch\n", argp[-2]);
					goto leave;
				}
				if(*folder == '+')
					folder++;
				folder = path(folder, TFOLDER);
				continue;
			case 6:
				if(filec >= NFOLD) {
					fprintf(stderr, "Too many src files.\n");
					goto leave;
				}
				if(!(cp = *argp++) || *cp == '-')
					goto missing;
				files[filec++] = path(cp, TFILE);
				continue;

							/* -help */
			case 7: help("file   [msgs] [switches]  +folder ...",
				     switches);
				goto leave;
			}
		if(*cp == '+')  {
			if(foldp < NFOLD)
				folders[foldp++].f_name = path(++cp, TFOLDER);
			else {
				fprintf(stderr, "Only %d folders allowed.\n", NFOLD);
				goto leave;
			}
		} else
			msgs[msgp++] = cp;
	}
	if(!m_find("path")) free(path("./", TFOLDER));
	if(!foldp) {
		fprintf(stderr, "No folder specified.\n");
fprintf(stderr, "Usage: file [-src folder] [msg ...] [switches] +folder [+folder]\n");
		goto leave;
	}
	if(filec > 1) {
		if(msgp) {
			fprintf(stderr, "File: Can't mix files and messages.\n");
			goto leave;
		}
		if(opnfolds())
			goto leave;
		for(i = 1; i < filec; i++) {
			if(m_file(folder, files[i], folders, foldp, prsrvf, 0))
				goto leave;
		}
		if(!linkf) {
			if((cp = m_find("delete-prog")) != NULL) {
				files[0] = r1bindex(cp, '/');
				execvp(cp, files);
				fprintf(stderr, "Can't exec deletion-prog--");
				perror(cp);
			} else for(i = 1; i < filec; i++) {
				if(unlink(files[i]) == -1) {
					fprintf(stderr, "Can't unlink ");
					perror(files[i]);
				}
			}
		}
		goto leave;
	}
	if(!msgp)
		msgs[msgp++] = "cur";
	if(!folder)
		folder = m_getfolder();
	VOID copy(m_maildir(folder), maildir);
	if(chdir(maildir) < 0) {
		fprintf(stderr, "Can't chdir to: ");
		perror(maildir);
		goto leave;
	}
	if(!(mp = m_gmsg(folder))) {
		fprintf(stderr, "Can't read folder %s!?\n",folder);
		goto leave;
	}
	if(mp->hghmsg == 0) {
		fprintf(stderr, "No messages in \"%s\".\n", folder);
		goto leave;
	}
	for(msgnum = 0; msgnum < msgp; msgnum++)
		if(!m_convert((cp = msgs[msgnum])))
			goto leave;
	if(mp->numsel == 0) {
		fprintf(stderr, "file:  ham 'n cheese\n");       /* never get here */
		goto leave;
	}
	m_replace(pfolder, folder);
	if(mp->hghsel != mp->curmsg && ((mp->numsel != mp->nummsg) || linkf))
		m_setcur(mp->hghsel);
	if(opnfolds())
		goto leave;
	for(msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
		if(mp->msgstats[msgnum] & SELECTED)
			if(m_file(folder, cp = getcpy(m_name(msgnum)),
				  folders, foldp, prsrvf, 0))
				goto leave;
			else
				cndfree(cp);
	if(!linkf) {
		if((cp = m_find("delete-prog")) != NULL) {
			if(mp->numsel > MAXARGS-2) {
	  fprintf(stderr, "file: more than %d messages for deletion-prog\n",MAXARGS-2);
				printf("[messages not unlinked]\n");
				goto leave;
			}
			vecp = 1;
			vec = (char **) calloc(MAXARGS + 2, sizeof *vec);
			for(msgnum= mp->lowsel; msgnum<= mp->hghsel; msgnum++)
				if(mp->msgstats[msgnum]&SELECTED)
					vec[vecp++] = getcpy(m_name(msgnum));
			vec[vecp] = 0;
			m_update();
			VOID fflush(stdout);
			vec[0] = r1bindex(cp, '/');
			execv(vec[0], vec);
			fprintf(stderr, "Can't exec deletion-prog--");
			perror(cp);
		} else {
			for(msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
				if(mp->msgstats[msgnum] & SELECTED)
					if(unlink(cp = m_name(msgnum))== -1) {
						fprintf(stderr, "Can't unlink %s:",folder);
						perror(cp);
					}
		}
	}
leave:
	m_update();
	done(0);
}


opnfolds()
{
	register int i;
	register char *cp;
	char nmaildir[128];

	for(i = 0; i < foldp; i++) {
		VOID copy(m_maildir(folders[i].f_name), nmaildir);
		if(access(nmaildir, 5) < 0) {
			cp = concat("Create folder \"", nmaildir, "\"? ", 0);
			if(!gans(cp, anoyes))
				goto bad;
			free(cp);
			if(!makedir(nmaildir)) {
				fprintf(stderr, "Can't create folder.\n");
				goto bad;
			}
		}
		if(chdir(nmaildir) < 0) {
			fprintf(stderr, "Can't chdir to: ");
			perror(nmaildir);
			goto bad;
		}
		if(!(folders[i].f_mp = m_gmsg(folders[i].f_name))) {
			fprintf(stderr, "Can't read folder %s\n", folders[i].f_name);
			goto bad;
		}
	}
	VOID chdir(maildir);    /* return to src folder */
	return(0);
bad:
	retur